pkg/kubelet/volumemanager/metrics/metrics.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/kubernetes/pkg/kubelet/volumemanager/cache"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const pluginNameNotAvailable = "N/A"
+
+var (
```

Comment:
Why are these global variables? Unless there is a good reason to deviate, follow the following pattern for new packages:

```GO
// Public interface exposed by this package.
type MyType interface {
  // you can embed another interface to require implementers to also implement other interfaces 
  prometheus.Collector

  // Required methods that other components will use to interact with this package.
  SomeInterfaceMethod()
}

// Public constructor that returns interface -- external components should only consume your interface not your concrete implementation.
func NewMyType(...) MyType {
}

// Private struct capturing any state that you need to persist for the lifecycle of the object
type myType struct {
}

// Compile-time check to ensure myType implements the MyType interface
var _ MyType = myType{}

func (t *myType) SomeInterfaceMethod() {
}
```

This makes it easy to mock out this package in testing and provides a clear (non-leaky) abstraction.

Reply:
If you need more help structuring this, let me know.

Reply:
From what I've seen, metrics variables are usually defined globally.

For example, this is the approache followed in Kubelet [1], Scheduler [2] and in Prometheus [3].

I'm not sure about the reason for that, but I found appropriate to follow the same pattern.

[1] https://github.com/bertinatto/kubernetes/blob/a160fe94a5b6f27a88761ecc9c2d9aeb60bb774a/pkg/kubelet/metrics/metrics.go#L67
[2] https://github.com/bertinatto/kubernetes/blob/a160fe94a5b6f27a88761ecc9c2d9aeb60bb774a/pkg/scheduler/metrics/metrics.go#L48
[3] https://github.com/prometheus/prometheus/blob/c28cc5076cd48c344d8c7b86a9c551f6c87b1fdf/storage/remote/queue_manager.go#L53

Reply:
It's worth thinking about the pros and cons of the pattern. There are a number of reasons to avoid global variables.

In general if we have a very clear input/output define for a piece of code with clear interfaces, it makes the code much easier to test and maintain (callers can easily clip out or replace dependencies). One argument I can buy (@gnufied pointed out to me offline), is that reinitialization of these variables is a bad enough that you want to programatically prevent that from happening, and package variables allow you to do that easily. I think even that can be accomplished without having to make everything a package variable.

That said, it's not a huge deal. I'm ok approving this as is.